carbonfive/spraygun-react-ts
# Upgrade to react-scripts 5

Problem
=======

The CRA team released [react-scripts 5.0.0](https://github.com/facebook/create-react-app/releases/tag/v5.0.0) recently. Depfu opened #242 to address this, however, react-scripts is responsible for generating many of the files in this repo, so PR does the upgrade manually to capture those changes.


Solution
========

Create a new project using react-scrips 5.0.0, audit the differences in the generated files from those currently on `main` (which were generated by an older version of react-scripts). Copy over those changes and try to get rid of some of the cruft in this template (there wasn't much).

Change summary
--------------

* Run create-react-app (with react-scripts 5.0.0) to generate a new, clean `app-prototype`
* Copy over our changes like SCSS, linter configs, docker config, CCI config, etc.
* Upgrade a handful of other packages that were lagging behind.

TODO
---------------------------------
- [x] Fix docker build (`$ docker build . -t spraygun-react-ts:latest`, [see comment](https://github.com/carbonfive/spraygun-react-ts/pull/249#issuecomment-1013585719))

Things to manually verify:
- [x] `yarn lint`
- [x] `yarn build`, `serve`
- [x] `yarn test`
- [x] `yarn start`
- [x] husky pre-commit hook
- [x] prettier (`eslint` works as expected)
- [x] Spraygun can use this template as expected (I generated this repo: https://github.com/dylnclrk/test-blog)
- [x] Circle CI configuration (tests pass on this branch)
- [x] Docker configuration (can run `docker build . -t spraygun-react-ts:latest && docker run -p 3000:3000 spraygun-react-ts:latest`)
- [x] Deploy to heroku (https://test-blog-tessssst.herokuapp.com)

Steps to Verify
---------------

1. Create a spraygun project with this version of the TS template (by using spraygun's  `-p path/to/this/template` flag)
2. See that it functions as expected (`yarn test`, `lint`, `start`, `build`, `yarn serve -s build` etc.)
3. Check that deploys to Heroku and runs on CCI as expected


New Dependencies
----------------

Additions:
* [web-vitals](https://snyk.io/advisor/npm-package/web-vitals) - comes with new CRA-generated apps, left it in.

🙈 See the package.json diff for all of the upgraded dependencies.



Screenshots
-----------

Appears to work locally:
<img width="500" alt="console with 'yarn lint' results, no warnings" src="https://user-images.githubusercontent.com/896684/149602550-05921f88-5e00-4d4a-a234-54353f86c571.png">
<img width="500" alt="console with 'yarn test' results, no failures" src="https://user-images.githubusercontent.com/896684/149602565-8b32dfb5-34e8-4d0b-8b7f-452146b50542.png">

<img width="1178" alt="app running successfully in browser" src="https://user-images.githubusercontent.com/896684/149602524-ad712822-631c-4306-8988-8dbb3568cc16.png">
